src/main/java/io/cdap/plugin/batch/sink/RecordToBSONWritableTransformer.java

```diff
@@ -0,0 +1,255 @@
+import java.math.BigInteger;
+import java.nio.ByteBuffer;
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * Transforms {@link StructuredRecord} to {@link BSONWritable}.
+ */
+public class RecordToBSONWritableTransformer {
+
+  private String idFieldName;
+
+  public RecordToBSONWritableTransformer() {
+  }
+
+  /**
+   * @param idFieldName specifies which of the incoming fields should be used as an document identifier. Identifier will
+   *                    be generated if no value is specified.
+   */
+  public RecordToBSONWritableTransformer(@Nullable String idFieldName) {
+    this.idFieldName = idFieldName;
+  }
+
+  /**
+   * Transforms given {@link StructuredRecord} to {@link BSONWritable}.
+   *
+   * @param record structured record to be transformed.
+   * @return {@link BSONWritable} that corresponds to the given {@link StructuredRecord}.
+   */
+  public BSONWritable transform(StructuredRecord record) {
+    return new BSONWritable(extractRecord(null, record, idFieldName));
+  }
+
+  private DBObject extractRecord(@Nullable String fieldName, @Nullable StructuredRecord record,
+                                 @Nullable String idFieldName) {
+    if (record == null) {
+      // Return 'null' value as it is
+      return null;
+    }
+    List<Schema.Field> fields = Objects.requireNonNull(record.getSchema().getFields(), "Schema fields cannot be empty");
+    BasicDBObjectBuilder bsonBuilder = BasicDBObjectBuilder.start();
+    for (Schema.Field field : fields) {
+      // Use full field name for nested records to construct meaningful errors messages.
+      // Full field names will be in the following format: 'record_field_name.nested_record_field_name'
+      String fullFieldName = fieldName != null ? String.format("%s.%s", fieldName, field.getName()) : field.getName();
+      Schema nonNullableSchema = field.getSchema().isNullable() ? field.getSchema().getNonNullable()
+        : field.getSchema();
+
+      if (field.getName().equals(idFieldName)) {
+        ObjectId objectId = extractObjectId(fullFieldName, record.get(field.getName()), nonNullableSchema);
+        bsonBuilder.add(MongoDBConstants.DEFAULT_ID_FIELD_NAME, objectId);
+        continue;
+      }
+
+      bsonBuilder.add(field.getName(), extractValue(fullFieldName, record.get(field.getName()), nonNullableSchema));
+    }
+    return bsonBuilder.get();
+  }
+
+  private ObjectId extractObjectId(String fieldName, Object value, Schema schema) {
+    Schema.Type type = schema.getType();
+    try {
+      if (type == Schema.Type.BYTES) {
+        return value instanceof ByteBuffer ? new ObjectId(((ByteBuffer) value).array()) : new ObjectId((byte[]) value);
```

Comment:
should not be using array(). You can use Bytes.toBytes(ByteBuffer)

Reply:
I was using `ByteBuffer#array` because of empty bytes array returned by `Bytes.toBytes(ByteBuffer)`. As it turned out, the root cause of this is a bug in the Source's `BSONObjectToRecordTransformer`. 

Added `ByteBuffer#rewind` method call to set ByteBuffer's position to 0. 
The line above changed to use